[benchmark] Eliminate usages of String.characters and CharacterView

moiseev opened the pull request:
To avoid compiler warnings.
<rdar://problem/34750654>

Commit:
[benchmark] Eliminate usages of String.characters and CharacterView

To avoid compiler warnings.
<rdar://problem/34750654>

moiseev commented:
@swift-ci Please smoke test

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
I had to modify the StringEdits benchmark. @swiftix, @atrick, @aschwaighofer is that OK, or should I go back to using `Array`?

Commit:
Apply changes to a gyb file instead of a generated result

moiseev commented:
@swift-ci Please smoke test

Commit:
Go back to using Array in StringEdits benchmark

moiseev commented:
@swift-ci Please smoke test

moiseev commented:
@swift-ci Please smoke benchmark

airspeedswift commented:
stringedits change lgtm (the `[CharacterView]` was essentially serving as a `[Substring]` anyway). Be interested to see if this makes any difference. Also, since `Substring` is hashable, would be interesting to know if using a `Set<Substring>` and eliminating the `map` at the end makes it faster or slower.

airspeedswift commented:
D'oh, sorry you already have the benchmark results. Hum, it made it a lot slower. That deserves investigating.

moiseev commented:
@airspeedswift well, of course it will make it slower. We're talking about adding to a set instead of adding to an array.

airspeedswift commented:
Yes but then it has to turn it into a Set at the end of the func. The q is, is it faster to build a giant array and dedupe at the end, vs build a set as you go. Build as you go does less work so should be faster, question is, is it.

> On Oct 3, 2017, at 9:49 AM, Max Moiseev <notifications@github.com> wrote:
> 
> @airspeedswift <https://github.com/airspeedswift> well, of course it will make it slower. We're talking about adding to a set instead of adding to an array.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/12196#issuecomment-333906640>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHr1paGfdTrgIKkt7nx7r_60dERkvXhoks5somWAgaJpZM4PpXaU>.
>

moiseev commented:
The second run of benchmarks is quite interesting. Even going back to using Array, but operating on Substring instead of CharacterView, makes it 110% slower.

moiseev commented:
@swift-ci Please smoke test

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
@swift-ci Please clean smoke test Linux platform

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
@swift-ci Please smoke benchmark

Commit:
Improve StringEdits performance in absence of CharacterView

CharacterView was not Hashable, so Set could not be used as an
accumulator. String and Substring are Hashable, but using a Set as an
accumulator is still slower than first collecting all the results in an
Array and then transforming it to a Set at the end. One possible reason
why that could be the case, is that by the time conversion happens, we
already know the capacity and thus will not re-allocate and re-hash a
Set on every insertion beyong current capacity.

moiseev commented:
@swift-ci Please smoke benchmark

moiseev commented:
@swift-ci Please smoke test

moiseev commented:
`StringEdits`  is almost back to norm.

moiseev commented:
@swiftix, @eeckstein, @airspeedswift et. al. Good to merge?

swiftix commented:
LGTM from me

eeckstein commented:
lgtm, too

palimondo commented on benchmark/single-source/StringEdits.swift:
```diff
@@ -31,14 +31,11 @@ var editWords: [String] = [
-  // create right/left splits as CharacterViews instead
   let splits = word.indices.map {
-    (word[..<$0], word[$0...])
+    (String(word[..<$0]), String(word[$0...]))
```
@airspeedswift @moiseev It seems to me that this fundamentally changed the core type in this benchmark, as it no longer operates on `Substring` views, but creates `String` copies, as can be seen in the modified type of `result` below.

This also implicitly changes the argument types of the string concatenation functions in the main workload.

Even with this change, the last benchmark run here has shown a slight regression:

TEST	| OLD	| NEW	| DELTA	| SPEEDUP
--|--|--|--|--|
StringEdits	| 122997	| 140316	| +14.1%	| 0.88x

and @moiseev said:
> `StringEdits` is almost back to norm.

…but this isn't the same benchmark anymore. When I've switched locally to running with `Substring`s, it get's worse (from: 264 500 to 293 400), but we are now running on different string guts, so its not directly comparable.

What should this benchmark test: `String` or `Substring`? 
Should I add a second variant for comparison?

moiseev replied:
> as it no longer operates on Substring views, but creates String copies

Because that's what it used to do before `Substring` was introduced. This change IIRC was made precisely because this was the minimal change to the benchmarks that would still allow us compare apples to apples, as opposed to rewriting a benchmark completely to use `Substring` as a common currency type.

I believe there is value in having both `String` and `Substring` based benchmarks, but @milseman would know for sure.

palimondo replied:
> Because that's what it used to do before Substring was introduced. 

If you look at the code — [[View Changes]](https://github.com/apple/swift/pull/12196/files/53ec0b1168fade46f83a478f0f12780ce7e7e710#diff-b251a3b7fa95677d2852e04f4097a2d9L41) — the original type was `[String.CharacterView]` which I believe corresponds to `Substring`, right?

palimondo replied:
Ping @milseman

palimondo replied:
Ping @milseman, @airspeedswift
